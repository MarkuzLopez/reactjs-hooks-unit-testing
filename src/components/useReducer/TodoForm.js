import React from 'react'
import { useForm } from '../useEffect/useForm';

export const TodoForm = (props) => {
    
    const { hadleAddTodo } = props;
    
    const [{ description }, hadleInputChange, resetForm]  = useForm({
        description: ''
    });

    const onSubmitTodo = (e) => {
        e.preventDefault();

        if(!description.trim().length){ return; }

        const newTodo = { 
            id: new Date().getTime(),
            desc: description,
            done: false   
        }

        hadleAddTodo(newTodo);
        resetForm();
    }

    
    return (
        <>
            <h1>TodoApp</h1>
            <hr />
            <form onSubmit={onSubmitTodo}>
                <input
                    type="text"
                    name="description"
                    placeholder="Aprener..."
                    autoComplete="off"
                    className="form-control"
                    value={description}
                    onChange={hadleInputChange}
                />

                <button
                    type="submit" className="btn btn-primary btn-lg btn-block mt-1">
                    Agreggar
                        </button>
            </form>
        </>
    )
}
