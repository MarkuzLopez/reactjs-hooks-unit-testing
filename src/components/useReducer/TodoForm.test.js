import { shallow } from "enzyme"
import { TodoForm } from "./TodoForm"

describe('pruebas todo form',  () => {
    
    const hadleAddTodo = jest.fn();

    const wrapper = shallow(
        <TodoForm
            hadleAddTodo={hadleAddTodo}
        />
    );

    test('debe de mostrarse correctamente ', () => {
        expect(wrapper).toMatchSnapshot();
    });

    test('NO debe de llamar hadleAddTodo sin valores', () => {
        
        const formSubmit = wrapper.find('form').prop('onSubmit');
        
        formSubmit({ 
            preventDefault(){ }
        });

        expect(hadleAddTodo).toHaveBeenCalledTimes(0);

    });

    test('debe de llamar la funcion hadleAddTodo con valores8', () => {
        
        const value = 'Aprende React';

        wrapper.find('input').simulate('change', {
            target: {
                value,
                name: 'description'
            }
        });

        const formSubmit = wrapper.find('form').prop('onSubmit');
        formSubmit({ preventDefault(){} });
        
        expect(hadleAddTodo).toHaveBeenCalledTimes(1);

        expect(hadleAddTodo).toHaveBeenCalledWith(expect.any(Object) );
        expect(hadleAddTodo).toHaveBeenCalledWith({
            id: expect.any(Number), // no me importa su valor solo que sea numero
            desc: value,
            done: false
        });

        // se ejecuta el resetForm
        expect(wrapper.find('input').prop('value')).toBe('');
    })
    
    
    
})
