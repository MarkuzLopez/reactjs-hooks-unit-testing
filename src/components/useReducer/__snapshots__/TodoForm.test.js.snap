// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pruebas todo form debe de mostrarse correctamente  1`] = `
<Fragment>
  <h1>
    TodoApp
  </h1>
  <hr />
  <form
    onSubmit={[Function]}
  >
    <input
      autoComplete="off"
      className="form-control"
      name="description"
      onChange={[Function]}
      placeholder="Aprener..."
      type="text"
      value=""
    />
    <button
      className="btn btn-primary btn-lg btn-block mt-1"
      type="submit"
    >
      Agreggar
    </button>
  </form>
</Fragment>
`;
